networkx/utils/backends.py

```diff
@@ -2491,6 +2513,12 @@ def _get_from_cache(cache, key, *, backend_name=None, mutations=None):
                 # the call to this dispatchable function will mutate an input.
                 mutations.append((cache, (ekey, nkey)))
             return (ekey, nkey), graph
+    if (rv := cache.get(key)) is not None:
+        # Return FAILED_TO_CONVERT when a conversion previously failed for
+        # this exact cache key. Instead of an exact key match, we could
+        # perhaps return FAILED_TO_CONVERT if any cache key dominated by
+        # the given cache key has previously failed to convert.
```

Comment:
This comment suggests there might be a more aggressive approach of returning `FAILED_TO_CONVERT` if any cache key "dominated by the given cache key" has failed.  How would we determine "dominated by"?  Is this just a matter of reverting line 2488 to not consider `rv != FAILED_TO_CONVERT`? Or is that loop not the same as "dominated by"?

Reply:
This is a great question and gets to the heart of understanding caching. Let me describe this with an example.

Say we want to convert a networkx graph to a backend graph and only want to preserve the "weight" edge attribute. We look for an already-converted graph in the cache that satisfies this criteria. This means we can used a cached graph that preserves "weight", or "weight" and "foo", or all attributes, etc. Similarly, when we save a converted graph to the cache, we remove previously-converted graphs that are no longer necessary. For example, if we converted while preserving "weight" and "foo" edge attributes, we can delete the conversion that only preserved the edge weight "weight".

So, how should we add "caching failed; do not try to convert" behavior to such a system where the only information we get is whether a conversion for a given cache key failed or not (for example, conversions from backends can't yet indicate "this failed because we couldn't convert 'foo' attribute"). When a conversion fails, we add `FAILED_TO_CONVERT` to the cache for they key that it failed for. When `_get_from_cache` searches the cache for a compatible graph, it ignores entries with `FAILED_TO_CONVERT`. In the line you indicated, `if (rv := cache.get(key)) is not None:` is _only_ true when this exact key failed to convert.

So what else could we do as indicated by my comment? We could search the cache as done in `_set_to_cache` for other keys that have `FAILED_TO_CONVERT` that the current key "dominates". For example, if we want to preserve both "weight" and "foo" edge attributes, we shouldn't convert (with high probability) if we previously failed to convert either "weight" or "foo" attribute.

I'm happy to try to update this language to something that makes more sense to you (and presumably others too).

Reply:
Thanks -- and which direction is the domination? In your example, does "preserve both 'weight' and 'foo'" dominate "preserve 'foo'" or the other way around? Is it different depending on whether you are setting or getting the entry?

Reply:
Yes, I would say "preserve both 'weight' and 'foo'" dominates "preserve 'foo'". So, preserving _all_ edge and node attributes "dominates" all others, and preserving _no_ attributes is "dominated by" all others. When adding a new item to the cache, we delete all former entries that the new item "dominates". When getting an item from the cache, we can get any entry that "dominates" what we are after. Regarding `FAILED_TO_CONVERT`, an option is to not attempt to convert if any entry that the current key "dominates" has `FAILED_TO_CONVERT`. I can try to implement the latter.

Reply:
Does this wording correctly reflect that idea (we could add the word "dominates" with defn here too if that's better):
```suggestion
        # Return FAILED_TO_CONVERT when a conversion previously failed for
        # this exact cache key. Instead of an exact key match, we could
        # perhaps return FAILED_TO_CONVERT if any cache key that requires all
        # edge/node attributes of the given cache key has previously failed to convert.
```